### PR TITLE
Limit Client Connections

### PR DESCRIPTION
**Pull request by vazois**
This PR exposes a parameter to allow for limiting the number of  client connections

### CODE REVIEW

File: libs/host/Configuration/Options.cs [hunk 1]
```diff
@@ -339,6 +339,10 @@ internal sealed class Options
         [Option("maxthreads", Required = false, HelpText = "Maximum worker and completion threads in thread pool, 0 uses the system default.")]
         public int ThreadPoolMaxThreads { get; set; }
 
+        [IntRangeValidation(0, int.MaxValue)]
+        [Option("connection-limit", Required = false, Default = 0, HelpText = "Maximum number of client connections")]
+        public int ConnectionLimit { get; set; }
```
Comment (msft-paddy14) [inline libs/host/Configuration/Options.cs:344]: will this be inclusive of Gossip connections incase of cluster mode?

Comment (vazois) [reply libs/host/Configuration/Options.cs:344]: yes, you have to account for the internode connections, which means gossip, replication, failover and migration connections. The failover and migration are transient connections until its operation finishes. Gossip and replication and long-lasting connections.

File: libs/server/Servers/GarnetServerTcp.cs [hunk 2]
```diff
@@ -134,11 +136,11 @@ private unsafe bool HandleNewConnection(SocketAsyncEventArgs e)
             string remoteEndpointName = e.AcceptSocket.RemoteEndPoint?.ToString();
             logger?.LogDebug("Accepted TCP connection from {remoteEndpoint}", remoteEndpointName);
 
-
             ServerTcpNetworkHandler handler = null;
             if (activeHandlerCount >= 0)
             {
-                if (Interlocked.Increment(ref activeHandlerCount) > 0)
+                var currentActiveHandlerCount = Interlocked.Increment(ref activeHandlerCount);
+                if (currentActiveHandlerCount is > 0 && (connectionLimit == 0 || currentActiveHandlerCount <= connectionLimit))
```
Comment (badrishc) [inline libs/server/Servers/GarnetServerTcp.cs:143]: no need for `is` for integers.

File: libs/server/Servers/GarnetServerOptions.cs [hunk 3]
```diff
@@ -246,6 +246,11 @@ public class GarnetServerOptions : ServerOptions
         /// </summary>
         public int ThreadPoolMaxThreads = 0;
 
+        /// <summary>
+        /// Maximum client connection limit
+        /// </summary>
+        public int ConnectionLimit = 0;
```
Comment (badrishc) [inline libs/server/Servers/GarnetServerOptions.cs:252]: We should let `0` denote "no connections allowed" which might be useful if for some reason people want to run the server without allowing any network connections. `-1` or `int.MaxValue` can be the default to allow unlimited connections.

File: libs/host/Configuration/Options.cs [hunk 4]
```diff
@@ -339,6 +339,10 @@ internal sealed class Options
         [Option("maxthreads", Required = false, HelpText = "Maximum worker and completion threads in thread pool, 0 uses the system default.")]
         public int ThreadPoolMaxThreads { get; set; }
 
+        [IntRangeValidation(0, int.MaxValue)]
+        [Option("connection-limit", Required = false, Default = 0, HelpText = "Maximum number of client connections")]
```
Comment (badrishc) [inline libs/host/Configuration/Options.cs:343]: Reword: `Maximum number of simultaneously active client connections.` (note also: added period at the end)

Comment (badrishc) [inline libs/host/Configuration/Options.cs:343]: Rename to `network-connection-limit` and `NetworkConnectionLimit` to clarify this does not apply to API-based session accesses.

File: libs/host/defaults.conf [hunk 5]
```diff
@@ -251,6 +251,9 @@
 	/* Maximum worker and completion threads in thread pool, 0 uses the system default. */
 	"ThreadPoolMaxThreads" : 0,
 
+	/* Maximum client connection limit. */
```
Comment (badrishc) [inline libs/host/defaults.conf:254]: Adjust to be consistent with the other two comment locations.

**Review by badrishc — APPROVED** [review]

**Pull request merged**